File: agents/BaseAgent.ts
// ============================================================
// BASE AGENT
// Abstract base class all agents extend. Handles Gemini calls,
// memory management, and RAG context injection.
// ============================================================

import { GoogleGenAI } from '@google/genai';
import { AgentModel, AgentStatus, AgentMemory } from '../models/AgentModel';
import { RetrievalResult } from '../rag/ragEngine';

const apiKey = import.meta.env.VITE_GEMINI_API_KEY || '';
export const geminiClient: GoogleGenAI | null =
  apiKey && apiKey !== 'YOUR_API_KEY_HERE'
    ? new GoogleGenAI({ apiKey })
    : null;

export const GEMINI_MODEL = 'gemini-1.5-flash-latest';

export abstract class BaseAgent {
  protected model: AgentModel;

  constructor(model: AgentModel) {
    this.model = model;
  }

  get id()     { return this.model.id; }
  get name()   { return this.model.name; }
  get status() { return this.model.status; }

  setStatus(status: AgentStatus) {
    this.model.status = status;
  }

  setTask(task: string) {
    this.model.currentTask = task;
  }

  addLog(entry: string) {
    const ts = new Date().toLocaleTimeString();
    this.model.logs = [`[${ts}] ${entry}`, ...this.model.logs].slice(0, 50);
    this.model.memory.shortTerm = [entry, ...this.model.memory.shortTerm].slice(0, 10);
  }

  learnPattern(pattern: string) {
    if (!this.model.memory.longTerm.includes(pattern)) {
      this.model.memory.longTerm = [pattern, ...this.model.memory.longTerm].slice(0, 30);
    }
  }

  getSnapshot(): AgentModel {
    return { ...this.model };
  }

  /** Build a full prompt = system prompt + training data + RAG context + user message */
  protected buildPrompt(userMessage: string, rag: RetrievalResult): string {
    return [
      this.model.systemPrompt,
      `\n\n--- AGENT TRAINING DATA ---\n${this.model.trainingData}\n---\n`,
      rag.context,
      `\n\nPrevious actions (short-term memory):\n${this.model.memory.shortTerm.slice(0, 5).join('\n') || 'None'}`,
      `\n\nLearned patterns (long-term memory):\n${this.model.memory.longTerm.slice(0, 5).join('\n') || 'None'}`,
      `\n\nCURRENT TASK:\n${userMessage}`
    ].join('');
  }

  /** Call Gemini with the full grounded prompt */
  protected async callGemini(prompt: string, temperature = 0.7): Promise<string> {
    if (!geminiClient) {
      throw new Error('Gemini API not configured. Set VITE_GEMINI_API_KEY in .env.local');
    }
    const response = await geminiClient.models.generateContent({
      model: GEMINI_MODEL,
      contents: prompt,
      config: { temperature, maxOutputTokens: 600 }
    });
    return response.text || '';
  }

  /** Parse JSON safely from Gemini response */
  protected parseJSON<T>(text: string, fallback: T): T {
    const match = text.match(/\{[\s\S]*\}/);
    if (match) {
      try { return JSON.parse(match[0]) as T; } catch { /* fall through */ }
    }
    return fallback;
  }
}

File: backend/agents/base_agent.py
"""
BaseAgent — all agents inherit from this.
Uses google-genai SDK with Gemini 3 Flash for agents,
Gemini 3.1 Pro Preview for the orchestrator.
"""
from __future__ import annotations
import json, os, time
from google import genai
from google.genai import types
from dotenv import load_dotenv
from backend.models.schemas import AgentModel, AgentStatus, AgentMemory
from backend.rag.rag_engine import RetrievalResult

load_dotenv()

# ── Gemini 3 models ───────────────────────────────────────────
AGENT_MODEL       = "gemini-3-flash"          # fast agents
ORCHESTRATOR_MODEL = "gemini-3.1-pro-preview"  # deep reasoning

# ── Singleton Gemini client ───────────────────────────────────
_gemini_client: genai.Client | None = None


def get_client() -> genai.Client:
    global _gemini_client
    if _gemini_client is None:
        api_key = os.getenv("GEMINI_API_KEY")
        if not api_key:
            raise RuntimeError("GEMINI_API_KEY not set in environment")
        _gemini_client = genai.Client(api_key=api_key)
    return _gemini_client


class BaseAgent:
    """Abstract base — subclasses implement execute()."""

    def __init__(self, model: AgentModel):
        self.model = model

    # ── Properties ────────────────────────────────────────────
    @property
    def id(self):     return self.model.id
    @property
    def name(self):   return self.model.name
    @property
    def status(self): return self.model.status

    # ── State helpers ─────────────────────────────────────────
    def set_status(self, status: AgentStatus):
        self.model.status = status

    def set_task(self, task: str):
        self.model.current_task = task

    def add_log(self, entry: str):
        ts = time.strftime("%H:%M:%S")
        log_line = f"[{ts}] {entry}"
        self.model.logs = [log_line] + self.model.logs[:49]
        self.model.memory.short_term = [entry] + self.model.memory.short_term[:9]

    def learn(self, pattern: str):
        if pattern not in self.model.memory.long_term:
            self.model.memory.long_term = [pattern] + self.model.memory.long_term[:29]

    def snapshot(self) -> dict:
        return self.model.model_dump()

    # ── Prompt builder ────────────────────────────────────────
    def _build_prompt(self, task: str, rag: RetrievalResult) -> str:
        memory_ctx = "\n".join(self.model.memory.short_term[:5]) or "None"
        learned    = "\n".join(self.model.memory.long_term[:5])  or "None"
        return (
            f"{self.model.system_prompt}\n\n"
            f"--- AGENT TRAINING DATA ---\n{self.model.training_data}\n---\n"
            f"{rag.context}"
            f"\nShort-term memory:\n{memory_ctx}"
            f"\nLearned patterns:\n{learned}"
            f"\n\nCURRENT TASK:\n{task}"
        )

    # ── Gemini call ───────────────────────────────────────────
    async def _call_gemini(
        self,
        prompt: str,
        temperature: float = 0.7,
        model: str = AGENT_MODEL,
    ) -> str:
        client = get_client()
        response = client.models.generate_content(
            model=model,
            contents=prompt,
            config=types.GenerateContentConfig(
                temperature=temperature,
                max_output_tokens=800,
            ),
        )
        return response.text or ""

    # ── JSON parser ───────────────────────────────────────────
    @staticmethod
    def _parse_json(text: str, fallback: dict) -> dict:
        import re
        match = re.search(r"\{[\s\S]*\}", text)
        if match:
            try:
                return json.loads(match.group())
            except json.JSONDecodeError:
                pass
        return fallback
